docs/examples/openai/openai_chat_with_hvplot.py

```diff
@@ -0,0 +1,261 @@
+PANEL_LOGO = {
+    "default": "https://panel.holoviz.org/_static/logo_horizontal_light_theme.png",
+    "dark": "https://panel.holoviz.org/_static/logo_horizontal_dark_theme.png",
+}
+PANEL_URL = "https://panel.holoviz.org/index.html"
+
+pn.chat.message.DEFAULT_AVATARS["assistant"] = HVPLOT_LOGO
+pn.chat.ChatMessage.show_reaction_icons = False
+
+
+@pn.cache
+def _read_data():
+    return (
+        pd.read_csv(ROOT / "ABC.csv", parse_dates=["date"])
+        .sort_values(by="date", ascending=False)
+        .head(100)
+    )
+
+
+DATA = _read_data()
+
+
+@pn.cache
+def _read_tool(name: str) -> dict:
+    # See https://json-schema.org/learn/glossary
+    with open(ROOT / f"tool_{name}.json", encoding="utf8") as file:
+        return json.load(file)
+
+
+def _set_theme():
+    if THEME == "dark":
+        pio.templates.default = "plotly_dark"
+        plt.style.use(["default", "dark_background"])
+    else:
+        pio.templates.default = "plotly"
+        plt.style.use(["default", "seaborn-v0_8"])
+
+
+TOOLS_MAP = {"hvplot": _read_tool("hvplot"), "renderer": _read_tool("renderer")}
+TOOLS = list(TOOLS_MAP.values())
+
+HVPLOT_ARGUMENTS = (
+    "`" + "`, `".join(TOOLS_MAP["hvplot"]["function"]["parameters"]["properties"]) + "`"
+)
+EXPLANATION = """
+## hvPlot by HoloViz
+---
+
+`hvPlot` is a high-level plotting library that that works almost in the same way as \
+the well known `Pandas` `.plot` method.
+
+The `.hvplot` method supports more data backends, plotting backends and provides more \
+features than the `.plot` method.
+
+## OpenAI GPT with Tools
+---
+
+We are using the OpenAI `{MODEL}` model with the `hvplot` and `renderer` *tools*.
+
+You can refer to the following `hvplot` arguments
+
+- {HVPLOT_ARGUMENTS}
+
+and `renderer` arguments
+
+- `backend`
+"""
+
+SYSTEM_PROMPT = """\
+You are now a **Plotting Assistant** that helps users plot their data using `hvPlot` \
+by `HoloViz`.\
+"""
+
+DATA_PROMPT = f"""\
+Hi. Here is a description of your `data`.
+
+The type is `{DATA.__class__.__name__}`. The `dtypes` are
+
+```bash
+{DATA.dtypes}
+```"""
+
+client = AsyncOpenAI()
+messages = [
+    {"role": "system", "content": SYSTEM_PROMPT},
+    {"role": "assistant", "content": DATA_PROMPT},
+]
+
+tools_pane = pn.pane.JSON(
+    object=TOOLS, depth=6, theme=JSON_THEME, name="Tools", sizing_mode="stretch_both"
+)
+tabs_layout = pn.Tabs(
+    pn.Column(name="Plot"),
+    tools_pane,
+    pn.Column(name="Arguments"),
+    sizing_mode="stretch_both",
+    styles={"border-left": "2px solid var(--neutral-fill-active)"},
+)
+
+
+def powered_by():
+    """Returns a component describing the frameworks powering the chat ui"""
+    params = {"height": 50, "sizing_mode": "fixed", "margin": (10, 10)}
+    return pn.Column(
+        pn.Row(
+            pn.pane.Image(CHAT_GPT_LOGO, **params),
+            pn.pane.Image(HVPLOT_LOGO, **params),
+        ),
+        sizing_mode="stretch_width",
+    )
+
+
+def to_code(kwargs):
```

Comment:
Wondering whether this should be marked as internal, e.g. `_to_code` for consistency?